scipy/_lib/_util.py

```diff
@@ -64,29 +94,29 @@ def _lazywhere(cond, arrays, f, fillvalue=None, f2=None):
     >>> _lazywhere(a > 2, (a, b), f, np.nan)
     array([ nan,  nan,  21.,  32.])
 
-    Notice, it assumes that all `arrays` are of the same shape, or can be
-    broadcasted together.
-
     """
-    cond = np.asarray(cond)
-    if fillvalue is None:
-        if f2 is None:
-            raise ValueError("One of (fillvalue, f2) must be given.")
-        else:
-            fillvalue = np.nan
+    xp = array_namespace(cond, *arrays)
+
+    if (f2 is fillvalue is None) or (f2 is not None and fillvalue is not None):
+        raise ValueError("Exactly one of `fillvalue` or `f2` must be given.")
+
+    args = xp.broadcast_arrays(cond, *arrays)
```

Comment:
While working on distribution infrastructure, I am considering overhead on the order of µs to be significant when (because users do notice). By that definition, `broadcast_arrays` has significant overhead.

```python3
import numpy as np
x = np.asarray([1, 2, 3, 4])
y = np.asarray([3, 4, 5, 6])

def f(*arrays):
    return np.broadcast_arrays(*arrays)

%timeit f(x, y)  # 2.64 µs ± 24.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

We can reduce that:

```python3
def f2(*arrays):
    arrays = [np.asarray(array) for array in arrays]
    shapes = {array.shape for array in arrays}
    return np.broadcast_arrays(*arrays) if len(shapes) > 1 else arrays
%timeit f2(x, y)  # 816 ns ± 1.36 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

# When broadcasting is actually necessary, the difference is a lot less noticeable
x = np.asarray(1)
y = np.asarray([3, 4, 5, 6])
%timeit f(x, y)  # 10.8 µs ± 82.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
%timeit f2(x, y)  # 11.9 µs ± 72.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Given that `_lazywhere` use is ubiquitous, would you mind if I implemented this optimization here?